src/lib/orion_websocket/parser.cpp

```diff
@@ -84,4 +87,42 @@ void ws_parser_parse
   {
     head.gotHeaders = false;
   }
+
+  if (doc.HasMember("params") && doc["params"].IsObject())
+  {
+    rapidjson::Value::ConstMemberIterator it = doc["params"].MemberBegin();
+    while (it != doc["params"].MemberEnd())
+    {
+      uriParamGet(ciP, it->name.GetString(), it->value.GetString());
+      ++it;
+    }
+  }
+}
+
+const char *ws_parser_message
+(
+ const std::string&  msg,
+ const HttpHeaders&  head,
+ int                 statusCode
+)
+{
+  const char *tmpl = "{\"headers\": %s, \"message\": %s, \"status\": \"%d\"}";
+
+  rapidjson::StringBuffer buff;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(buff);
+  writer.StartObject();
+  std::map<std::string, std::string *>::const_iterator it = head.headerMap.begin();
+  while (it != head.headerMap.end())
+  {
+    writer.Key(it->first.c_str());
+    writer.String(it->second->c_str());
+    ++it;
+  }
+  writer.EndObject();
+
+  const char *headers = buff.GetString();
+  char *json = (char *) malloc(strlen(tmpl) + strlen(headers) + msg.size() + 1);
+  sprintf(json, tmpl, headers, msg.c_str(), statusCode);
```

Comment:
The 'json' buffer is not big enough here - you may write over it, causing memory corruption.

More correct size:

    strlen(tmpl) - 6 + strlen(headers) + msg.size() + 10 + 1;

    // The '6' comes from strlen of %s + %s + %d (removed from total len of 'tmpl')
    // The '10' is the maximum number of bytes in a string a 32 bit integer may need.
    // The '1' is of course cor the null termination of the string.

Also, if you use snprintf instead of sprintf any possible memory corruption is avoided. The entire string may not get inside, but at least there will never be any memory corruption - a lot easier to troubleshoot ...